kedro/framework/cli/cli.py

```diff
@@ -107,15 +115,50 @@ class KedroCLI(CommandCollection):
 
     def __init__(self, project_path: Path):
         self._metadata = None  # running in package mode
+        self._guessed_project_root = None
+
         if _is_project(project_path):
             self._metadata = bootstrap_project(project_path)
+        else:
+            guessed_project_root, _ = find_project_root(None)
+            if _is_project(guessed_project_root):
+                self._guessed_project_root = guessed_project_root
         self._cli_hook_manager = get_cli_hook_manager()
 
         super().__init__(
             ("Global commands", self.global_groups),
             ("Project specific commands", self.project_groups),
         )
 
+    def get_command(self, ctx: Context, cmd_name: str) -> Optional[Command]:
```

Comment:
I'm a huge fan of this change in principle, thank you for adding it! However, please could you split it off into a separate PR? There's a few things I think need discussing:

* not at all sure about using `black` for `find_project_root`. Note we have something similar `_find_kedro_project` in the ipython extension; if that seems suitable to reuse, let's move it to some common utils file and use it here
* more importantly, if we are inside a kedro project, just not at the top level, is there actually any reason these project commands shouldn't work? 🤔 Instead of an error message telling the user to `cd` can we actually just somehow make kedro work inside a project subfolder (like `kedro ipython` does)? But without `cd`ing to the kedro project top-level
* if we're totally outside a kedro project, a helpful error message is great but:
  * is there some better way to check if the command is a project one without overriding `get_command`, e.g. check if it's in `self.project_groups`?
  * I don't think this formatting is "on brand" yet since we haven't moved the rest of the CLI to rich yet. Let's just do it as `click.secho` for now. Doing rich styling for CLI messages is a separate issue for another time (e.g. we might move to https://github.com/ewels/rich-click)

Reply:
#1720 Separate PR created here, I put it in the inbox so we can discuss it a bit more.